core/src/mast/mod.rs

```diff
@@ -0,0 +1,172 @@
+use alloc::{collections::BTreeMap, vec::Vec};
+use miden_crypto::hash::rpo::RpoDigest;
+
+use crate::{DecoratorList, Kernel, Operation};
+
+mod basic_block_node;
+pub use basic_block_node::BasicBlockNode;
+
+mod call_node;
+pub use call_node::CallNode;
+
+mod dyn_node;
+pub use dyn_node::DynNode;
+
+mod join_node;
+pub use join_node::JoinNode;
+
+mod split_node;
+pub use split_node::SplitNode;
+
+mod loop_node;
+pub use loop_node::LoopNode;
+
+pub trait MerkleTreeNode {
+    fn digest(&self) -> RpoDigest;
+}
+
+/// An opaque handle to a [`MastNode`] in some [`MastForest`]. It is the responsibility of the user
+/// to use a given [`MastNodeId`] with the corresponding [`MastForest`].
+///
+/// Note that since a [`MastForest`] enforces the invariant that equal [`MastNode`]s MUST have equal
+/// [`MastNodeId`]s, [`MastNodeId`] equality can be used to determine equality of the underlying
+/// [`MastNode`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct MastNodeId(usize);
```

Comment:
**Suggestion:** Make `MastNodeId` wrap a u32 value, rather than `usize`, with appropriate trait impls and helpers so as to make indexing with it still reasonably ergonomic.

Doing so will result in significant memory savings when many programs are cached in memory, and is more efficient to copy around. Another benefit is it puts a reasonable finite cap on the maximum number of MAST nodes in a single forest, ensuring that programs cannot consume so much memory that it exhausts what is available to the VM.

The main downside of course, is that it requires making allocation of a `MastNodeId` from a `MastForest` fallible, and limits the size of a `MastForest` (both a pro and a con).

We do the same thing in the compiler for all of our entity identifiers, as does Cranelift. LLVM tends to use pointers and uniquing (i.e. two entities that are equal, always have the same pointer value), which is another approach one can take, but is not as nice to work with in Rust due to lifetimes.

**Suggestion:** Derivfe `PartialOrd`, `Ord`, and `Hash` here, since those are well defined here.

Reply:
> The main downside of course, is that it requires making allocation of a MastNodeId from a MastForest fallible, and limits the size of a MastForest (both a pro and a con).

Yes I agree with the suggestion; we actually have the assumption that `usize` is at least as big as `u32` elsewhere in the codebase (and in winterfell). And one "limitation" of our field is that we can't prove traces longer than `2^32` - which translates to definitely never more than `2^32` `MastNode`s in a forest.

Will change it to `u32`.

Reply:
> And one "limitation" of our field is that we can't prove traces longer than `2^32` - which translates to definitely never more than `2^32` `MastNode`s in a forest.

Technically, we could have a program with more than $2^{32}$ MAST nodes but that would take less than $2^{32}$ cycles to execute. This could be a program with many branches where a small portion of the code is actually executed for any given input.

But I think this is a technicality and I think using `u32` for node index is fine.